### PR TITLE
Generate IEquatable support and operators when generating .Equals for structs

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/30396

### CODE REVIEW

File: src/EditorFeatures/VisualBasicTest/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.vb [hunk 1]
```diff
@@ -89,6 +89,38 @@ End Class",
 index:=1)
         End Function
 
+        <WorkItem(30396, "https://github.com/dotnet/roslyn/issues/30396")>
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateEqualsAndGetHashCode)>
+        Public Async Function TestStructure() As Task
```
Comment (sharwell) [inline src/EditorFeatures/VisualBasicTest/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.vb]: 💡 An additional test for already implementing `IEquatable(Of Z)` would be good

Comment (sharwell) [inline src/EditorFeatures/VisualBasicTest/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.vb]: 💡 An additional test for already implementing operators would be good

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs [hunk 2]
```diff
@@ -421,27 +421,199 @@ struct ReallyLongName
...
+
+    public static bool operator !=(ReallyLongName left, ReallyLongName right)
+    {
+        return !(left == right);
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateEqualsAndGetHashCode)]
+        public async Task TestEqualsStructAleadyHasOperators()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs]:
```suggestion
        public async Task TestEqualsStructAlreadyHasOperators()
```

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs [hunk 3]
```diff
@@ -421,27 +421,199 @@ struct ReallyLongName
...
+    {
+        return !(left == right);
+    }
+}");
         }
 
-        var name = (ReallyLongName)obj;
-        return i == name.i &&
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateEqualsAndGetHashCode)]
+        public async Task TestEqualsStructAleadyImplementsIEquatable()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs]:
```suggestion
        public async Task TestEqualsStructAlreadyImplementsIEquatable()
```

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs [hunk 4]
```diff
@@ -421,27 +421,199 @@ struct ReallyLongName
...
+
+struct ReallyLongName : IEquatable<ReallyLongName>
+{
+    int i;
+
+    string S { get; }
+
+    public override bool Equals(object obj)
+    {
+        return obj is ReallyLongName name && Equals(name);
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs]: ❔ Do we have a test for cases where `obj is ReallyLongName name` is not supported by the current language version?

File: src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -166,6 +132,21 @@ private bool HasOperator(INamedTypeSymbol containingType, string operatorName)
                                        containingType.Equals(m.Parameters[0].Type) &&
                                        containingType.Equals(m.Parameters[1].Type));
 
+        private bool CanImplementsIEquatable(
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs]: Done.  Thanks!

File: src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs [hunk 6]
```diff
@@ -223,48 +203,98 @@ private void GetExistingMemberInfo(INamedTypeSymbol containingType, out bool has
...
+        }
+
+        private async Task<CodeAction> CreateCodeActionWithDialogAsync(
+            Document document, TextSpan textSpan, SyntaxNode typeDeclaration, INamedTypeSymbol containingType, ImmutableArray<ISymbol> members,
+            bool generateEquals, bool generateGetHashCode, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            var pickMembersOptions = ArrayBuilder<PickMembersOption>.GetInstance();
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs]: Done.

File: src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs [hunk 7]
```diff
@@ -223,48 +203,98 @@ private void GetExistingMemberInfo(INamedTypeSymbol containingType, out bool has
...
+            var generateOperators = false;
+
+            if (generateEquals && containingType.TypeKind == TypeKind.Struct)
             {
-                return new GenerateEqualsAndGetHashCodeAction(
-                    document, textSpan, typeDeclaration, containingType, members,
-                    generateEquals, generateGetHashCode,
-                    implementIEquatable: false, generateOperators: false);
+                // if we're generating equals for a struct, then also add IEquatable<S> support as
+                // well as operators (as long as the struct does not already have them.
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs]: done.

File: src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs [hunk 8]
```diff
@@ -223,48 +203,98 @@ private void GetExistingMemberInfo(INamedTypeSymbol containingType, out bool has
...
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                implementIEquatable = CanImplementsIEquatable(semanticModel, containingType, out var equatableTypeOpt);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs]: done.

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs]: fixed all cases.

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs]: yup. `TestEqualsSingleField` does this by passing in C#6.  It validates you get:

```
    public override bool Equals(object obj)
    {
        var program = obj as Program;
        return program != null &&
               a == program.a;
    }
```

File: src/EditorFeatures/VisualBasicTest/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.vb (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasicTest/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.vb]: done.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasicTest/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.vb]: done.

File: src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs [hunk 9]
```diff
@@ -166,6 +132,21 @@ private bool HasOperator(INamedTypeSymbol containingType, string operatorName)
...
 
+        private bool CanImplementsIEquatable(
+            SemanticModel semanticModel, INamedTypeSymbol containingType,
+            [NotNullWhen(true)] out INamedTypeSymbol constructedType)
+        {
+            var equatableTypeOpt = semanticModel.Compilation.GetTypeByMetadataName(typeof(IEquatable<>).FullName);
+            if (equatableTypeOpt != null)
+            {
+                constructedType = equatableTypeOpt.Construct(containingType);
+                return !containingType.AllInterfaces.Contains(constructedType);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs]: I'm going to keep this logic for now absent any known issues :)

[the remark this answers is not in the document]
Comment (sharwell) [reply src/Features/Core/Portable/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersCodeRefactoringProvider.cs]: I'd pass `SymbolEqualityComparer.Default` so it's clear this is the comparison we want

File: src/EditorFeatures/Test/CodeFixes/CodeFixServiceTests.cs [hunk 10]
```diff
@@ -498,7 +498,6 @@ public async Task TestNuGetAndVsixCodeFixersWithMultipleFixableDiagnosticIdsAsyn
             return await fixService.GetFixesAsync(document, TextSpan.FromBounds(0, 0), includeConfigurationFixes: false, cancellationToken: CancellationToken.None);
         }
 
-        [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
```
Comment (sharwell) [inline src/EditorFeatures/Test/CodeFixes/CodeFixServiceTests.cs]: ❔ Why is this getting removed? I would have expected this:

```csharp
[ExportCodeFixProvider(LanguageNames.CSharp), Shared, PartNotDiscoverable]
```

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test/CodeFixes/CodeFixServiceTests.cs]: That's bizarre.  i don't remember making that change...

Comment (sharwell) [reply src/EditorFeatures/Test/CodeFixes/CodeFixServiceTests.cs]: Looks like it was pulled in from a separate change

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**